src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -0,0 +1,234 @@
+		ErrorIfCoordinatorInsertSelectUnsupported(insertSelectQuery);
+
+	if (multiPlan->planningError != NULL)
+	{
+		return multiPlan;
+	}
+
+	if (list_length(insertSelectQuery->cteList) > 0)
+	{
+		struct DecrementCTELevelsContext context;
+		List *combinedCTEList = NIL;
+
+		/* prepend top-level CTEs in subquery */
+		combinedCTEList = list_concat(combinedCTEList, insertSelectQuery->cteList);
+		combinedCTEList = list_concat(combinedCTEList, subquery->cteList);
+		subquery->cteList = combinedCTEList;
+
+		context.topLevelCTEList = insertSelectQuery->cteList;
+		context.topLevelsUp = 0;
+
+		/* correct ctelevelsup for range table entries that refer to top-level CTEs */
+		DecrementCTELevelsWalker((Node *) subquery, &context);
+	}
+
+	ReorderInsertSelectTargetLists(insertSelectQuery, insertRte, subqueryRte);
+
+	/* add casts when the SELECT output does not directly match the table */
+	forboth(insertTargetCell, insertSelectQuery->targetList,
+			selectTargetCell, subquery->targetList)
+	{
+		TargetEntry *insertTargetEntry = (TargetEntry *) lfirst(insertTargetCell);
+		TargetEntry *selectTargetEntry = (TargetEntry *) lfirst(selectTargetCell);
+
+		Var *columnVar = NULL;
+		Oid columnType = InvalidOid;
+		int32 columnTypeMod = 0;
+		Oid selectOutputType = InvalidOid;
+
+		/* indirection is not supported, e.g. INSERT INTO table (composite_column.x) */
+		if (!IsA(insertTargetEntry->expr, Var))
+		{
+			ereport(ERROR, (errmsg("can only handle regular columns in the target "
+								   "list")));
+		}
+
+		columnVar = (Var *) insertTargetEntry->expr;
+		columnType = get_atttype(targetRelationId, columnVar->varattno);
+		columnTypeMod = get_atttypmod(targetRelationId, columnVar->varattno);
+		selectOutputType = columnVar->vartype;
+
+		/*
+		 * If the type in the target list does not match the type of the column,
+		 * we need to cast to the column type. PostgreSQL would do this
+		 * automatically during the insert, but we're passing the SELECT
+		 * output directly to COPY.
+		 */
+		if (columnType != selectOutputType)
+		{
+			Expr *selectExpression = selectTargetEntry->expr;
+			Expr *typeCast = (Expr *) coerce_to_target_type(NULL,
+															(Node *) selectExpression,
+															selectOutputType,
+															columnType,
+															columnTypeMod,
+															COERCION_EXPLICIT,
+															COERCE_IMPLICIT_CAST,
+															-1);
+
+			selectTargetEntry->expr = typeCast;
+		}
+	}
+
+	multiPlan->insertSelectQuery = subquery;
+	multiPlan->insertTargetList = insertSelectQuery->targetList;
+	multiPlan->targetRelationId = targetRelationId;
+
+	return multiPlan;
+}
+
+
+/*
+ * ErrorIfCoordinatorInsertSelectUnsupported returns an error if executing an
+ * INSERT ... SELECT command by pulling results of the SELECT to the coordinator
+ * is unsupported because it uses RETURNING, ON CONFLICT, or an append-distributed
+ * table.
+ */
+static DeferredErrorMessage *
+ErrorIfCoordinatorInsertSelectUnsupported(Query *insertSelectQuery)
```

Comment:
Very minor: For subquery pushdown, we used the prefix `DeferErrorIf..`. But, pick whichever you want.